base/tps/src/org/dogtagpki/server/tps/processor/TPSEnrollProcessor.java

```diff
@@ -3472,27 +3505,41 @@ private RSAPublicKey parsePublicKeyBlob(
     public String getAESKeyWrapAlgSSKeyGen() {
 
         String aesKeyWrapAlg = "KWP";
-
-        IConfigStore configStore = CMS.getConfigStore();
-
-        // op.enroll.userKey.keyGen.aesKeyWrapAlg 
-        try {
-            String configValue = TPSEngine.OP_ENROLL_PREFIX + "." + selectedTokenType + "." + TPSEngine.CFG_KEYGEN +
-                   "." +  TPSEngine.CFG_AES_KEY_WRAP_ALG;
-
-            CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: configValue . " + configValue);
-            aesKeyWrapAlg = configStore.getString(
-                    configValue, "KWP");
-            CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: value " + aesKeyWrapAlg);
-
-        } catch (EBaseException e) {
-            //return default
-            return aesKeyWrapAlg;
+       
+        // Applet and Alg Selection by Token Range Support - check keyWrapAlg target configuration in token range first
+        String selectedAlg = getSelectedKeyWrapAlg();
+        if (selectedAlg == null) {  // Applet and Alg Selection by Token Range Support - use aesKeyWrapAlg configured by tokenType
+            IConfigStore configStore = CMS.getConfigStore();
+            
+            // op.enroll.userKey.keyGen.aesKeyWrapAlg 
+            try {
+                String configValue = TPSEngine.OP_ENROLL_PREFIX + "." + selectedTokenType + "." + TPSEngine.CFG_KEYGEN +
+                       "." +  TPSEngine.CFG_AES_KEY_WRAP_ALG;
+    
+                CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: configValue . " + configValue);
+                aesKeyWrapAlg = configStore.getString(
+                        configValue, "KWP");
+                CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: value " + aesKeyWrapAlg);
+                // Red Hat call into the new method to set DES to allow it to work for the non external reg case.
+                if(aesKeyWrapAlg.equalsIgnoreCase("DES")) {
+                    CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen DES configured per original token based value.");
+                    setSelectedKeyWrapAlg(aesKeyWrapAlg);
```

Comment:
just to echo what I mentioned earlier, the call to  setSelectedKeyWrapAlg(...) essentially makes the getAESKeyWrapAlgSSKeyGen() method a misnomer that ostensibly looks like a "get" method but is setting a global variable underneath.  It would be better to add the set* call after the get* call, not inside of it.

Reply:
The reason why I did it this was was to isolate the change inside of that method. It only calls setSelectedKeyWrapAlg in the DES case. If I do what you suggest, then every place getAESKeyWrapAlgSSKeyGen is called, we would have to explicitly call the setSelectedKeyWrapAlg. The goal was to change as little of the code as possible to get this working for the non external Reg case. Note that if setSelectedKeyWrapAlg has already been called by the code doing external reg (based on keyset mapping) this method just returns that value.

Here is the original version of their patch for this method:

+    public String getAESKeyWrapAlgSSKeyGen() {
+
+        String aesKeyWrapAlg = "KWP";
+        
+        // Applet and Alg Selection by Token Range Support - check keyWrapAlg target configuration in token range first
+        String selectedAlg = getSelectedKeyWrapAlg();
+        if (selectedAlg == null) {  // Applet and Alg Selection by Token Range Support - use aesKeyWrapAlg configured by tokenType
+            IConfigStore configStore = CMS.getConfigStore();
+            
+            // op.enroll.userKey.keyGen.aesKeyWrapAlg 
+            try {
+                String configValue = TPSEngine.OP_ENROLL_PREFIX + "." + selectedTokenType + "." + TPSEngine.CFG_KEYGEN +
+                       "." +  TPSEngine.CFG_AES_KEY_WRAP_ALG;
+    
+                CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: configValue . " + configValue);
+                aesKeyWrapAlg = configStore.getString(
+                        configValue, "KWP");
+                CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: value " + aesKeyWrapAlg);
+    
+            } catch (EBaseException e) {
+                //return default
+                return aesKeyWrapAlg;
+            }
+        } else if (!selectedAlg.equalsIgnoreCase("DES")) {
+            aesKeyWrapAlg = selectedAlg;
+            CMS.debug("TPSEnrollProcessor::getAESKeyWrapAlgSSKeyGen: using keyWrapAlg configured by token range: " + aesKeyWrapAlg);
+        }
+        
+        return aesKeyWrapAlg;
+
+    }

Reply:
Also found and removed one debug line with possibly objectionable data and re-pushed the PR.

Reply:
There are a total of 7 calls to getAESKeyWrapAlgSSKeyGen, all inside TPSEnrollProcessor.java.  Perhaps you could change setSelectedKeyWrapAlg() to call getAESKeyWrapAlgSSKeyGen instead? and call setSelectedKeyWrapAlg for those 7 calls to getAESKeyWrapAlgSSKeyGen instead.   In my humble opinion, having set* routine to call get* is fine, but changing get* to do set* isn't so much.  What do you think?

Reply:
or create a 3rd method to wrap it if it'd make better sense.

Reply:
Well what we have here is two different values representing the same thing.  They have already modified my original method to support their new value keyWrapAlg. Note in effect they are "setting" the return of my method to their value of keyWrapAlg. All I did was to do the reverse. The idea to call setSelectedKeyWrap from this code doesn't work because they are setting the value based on the config in the keyset mapping config. I think the only  way to do what you want is to just sprinkle calls to setSelectedKeyWrapAlg in those 7 places. Also in those 7 places I would have to check to see if the value keyWrapAlg has not already been set and if it's DES. Here its encapsulated inside a method. I will do what you think is best though. thanks.